src/backend/distributed/progress/multi_progress.c

```diff
@@ -0,0 +1,287 @@
+												  Datum argument);
+
+
+/*
+ * CreateProgressMonitor is used to create a place to store progress information related
+ * to long running processes. The function creates a dynamic memory segment consisting of
+ * a header regarding to the process and an array of "steps" that the long running
+ * "operations" consists of. The handle of the dynamic shared memory is stored in
+ * pg_stat_get_progress_info output, to be parsed by a progress retrieval command
+ * later on. This behavior may cause unrelated (but hopefully harmless) rows in
+ * pg_stat_progress_vacuum output. The caller of this function should provide a magic
+ * number, a unique 64 bit unsigned integer, to distinguish different types of commands.
+ */
+void *
+CreateProgressMonitor(uint64 progressTypeMagicNumber, int stepCount, Size stepSize,
+					  Oid relationId)
+{
+	dsm_segment *dsmSegment = NULL;
+	dsm_handle dsmHandle = 0;
+	void *dsmSegmentAddress = NULL;
+	ProgressMonitorHeader *header = NULL;
+	void *steps = NULL;
+	Size monitorSize = 0;
+
+	if (stepSize <= 0 || stepCount <= 0)
+	{
+		ereport(ERROR,
+				(errmsg("Number of steps and size of each step should be positive "
+						"values.")));
+	}
+
+	monitorSize = sizeof(ProgressMonitorHeader) + stepSize * stepCount;
+	dsmSegment = dsm_create(monitorSize, DSM_CREATE_NULL_IF_MAXSEGMENTS);
+
+	if (dsmSegment == NULL)
+	{
+		ereport(WARNING,
+				(errmsg(
+					 "Couldn't create a dynamic shared memory segment to keep track of "
+					 "progress of the current command.")));
+		return NULL;
+	}
+
+	dsmSegmentAddress = dsm_segment_address(dsmSegment);
+	dsmHandle = dsm_segment_handle(dsmSegment);
+
+	MonitorInfoFromAddress(dsmSegmentAddress, &header, &steps);
+
+	header->stepCount = stepCount;
+	header->processId = MyProcPid;
+
+	pgstat_progress_start_command(PROGRESS_COMMAND_VACUUM, relationId);
+	pgstat_progress_update_param(1, dsmHandle);
+	pgstat_progress_update_param(0, progressTypeMagicNumber);
+
+	currentProgressDSMHandle = dsmHandle;
+
+	return steps;
+}
+
+
+/*
+ * GetProgressMonitor function returns the header and steps array related to the current
+ * progress. A progress monitor should be created by calling CreateProgressMonitor,
+ * before calling this function.
+ */
+void
+GetProgressMonitor(ProgressMonitorHeader **header, void **steps)
+{
+	dsm_segment *dsmSegment = dsm_find_mapping(currentProgressDSMHandle);
+	void *monitorAddress = NULL;
+
+	if (dsmSegment == NULL)
+	{
+		dsmSegment = dsm_attach(currentProgressDSMHandle);
+	}
+
+	if (dsmSegment == NULL)
+	{
+		header = NULL;
+		steps = NULL;
+	}
+	else
+	{
+		monitorAddress = dsm_segment_address(dsmSegment);
+		MonitorInfoFromAddress(monitorAddress, header, steps);
+	}
+}
+
+
+/*
+ * FinalizeProgressMonitor releases the dynamic memory segment of the current progress
+ * and removes the process from pg_stat_get_progress_info() output.
+ */
+void
+FinalizeProgressMonitor(void)
```

Comment:
Changed names to `FinalizeCurrentProgressMonitor` and `GetCurrentProgressMonitor`